---
title: "Differential Expression"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Differential Expression}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, include = FALSE}
library(broadSeq)
```

```{r eval=FALSE}
library(broadSeq)
```

In this vignette, the easy use of broadSeq to identify differentially expressed(DE) genes with several R packages has been shown. All these packages use different input and output formats. Whereas with broadSeq, only `SummarizedExperiment` format will be enough for input. And the results will be returned as `data.frame` object which will make downstream analysis easier. No need to learn new data formats for evaluating different packages.

## Data 

In this example we will use RNA-seq expression data from developing mouse molar tissue. First we read the data as a `SummarizedExperiment` object. 

```{r warning=FALSE,eval=TRUE}
se <- readRDS(system.file("extdata","rat_vole_mouseSE_salmon.rds", package = "broadSeq"))

# To reduce the runtime, subset of the data used here
se <- se[rowData(se)$chromosome_name == 2,colData(se)$species == "Mouse"]
```


```{r warning=FALSE,eval=FALSE,echo=FALSE}
count_matrix <- as.matrix(read.table(file = system.file("extdata", 
                                              "tooth_RNASeq_counts.txt", 
                                              package = "DELocal")))
colData <- data.frame(condition=gsub("\\..*",x=colnames(count_matrix),replacement = ""))

gene_location <- read.table(file = system.file("extdata", 
                                              "gene_location.txt", 
                                              package = "DELocal"))

se <- SummarizedExperiment::SummarizedExperiment(assays=list(counts=count_matrix),
                                                      rowData = gene_location, 
                                                      colData=colData)
# contrast= c("condition","ME13","ME14")
```

### Gene information

```{r warning=FALSE}
head(rownames(se))
```

The gene ids are in ensemble format. For easy readability, information about genes can be fetched through `biomaRt` package as shown below.

<!-- Let us check first the available information about genes/feature   -->

```{r warning=FALSE,eval=FALSE}
require(biomaRt)

ensembl_ms_mart <- useMart(biomart="ENSEMBL_MART_ENSEMBL",
                           dataset="mmusculus_gene_ensembl", host="www.ensembl.org")
gene_attributes<- c("ensembl_gene_id", "mgi_symbol",
                    "start_position", "chromosome_name","gene_biotype")
gene_location_sample <- getBM(attributes=gene_attributes, filters = "ensembl_gene_id",
                              values = rownames(se),uniqueRows = TRUE,
                              mart=ensembl_ms_mart,verbose = FALSE,useCache = TRUE)

# remove rows with duplicate ensembl_gene_id
gene_location_sample <- gene_location_sample[!duplicated(gene_location_sample$ensembl_gene_id),]
# some ensembl_gene_id may not be found in biomart
# remove those ensembl_gene_id s
se <- se[gene_location_sample$ensembl_gene_id,]

rownames(gene_location_sample) <- gene_location_sample$ensembl_gene_id
rowData(se) <- gene_location_sample[rownames(se),]
```

Finally the gene information is stored and can be accessed through rowData funtion  

```{r warning=FALSE,eval=FALSE}
head(rowData(se))
```

### Sample information  

The sample metadata  

```{r warning=FALSE,eval=TRUE}
head(colData(se))
table(colData(se)$stage)
```
<!-- There are two biological conditions ME13 and ME14 (embryonic days 13 and 14) with seven replicates each. The relevant question is which genes are differentially expressed between these two developmental time points. -->

## Diff Expression  

There are four developmental time points; Bud, Cap, Late Cap and Bell (from embryonic days 13 to 16) with seven replicates each. Here, in this example, DE genes between Bud and Cap stages will be identified.

### Function pattern  

In broadSeq, the names of DE method has similar pattern like "use_"+"method name". All these method has same signature of input arguments. 

Here we will see `use_NOIseq()` method to apply `NOISeq::noiseqbio()` function.   

```{r warning=FALSE,fig.width=6,fig.height=6,message=FALSE}
result_Noiseq <- 
    use_NOIseq(se = se, 
           colData_id = "stage", control = "Bud", treatment = "Cap",
           rank = TRUE, 
           r = 10) # r is an argument of NOISeq::noiseqbio

head(result_Noiseq)
```



```{r warning=FALSE,fig.width=8,fig.height=6,message=FALSE}
# boxplot of top four DE genes with their mgi_symbol

# broadSeq::genes_plot function internally converts SummarizedExperiment to use
# ggpubr::ggboxplot 

pg <- broadSeq::genes_plot(se, x = "stage", assayName =  "counts",  
           features = result_Noiseq %>% dplyr::filter(rank <5) %>% rownames(),
           fill="stage", facet.by = "symbol",
           palette="jco", add = "dotplot")+rotate_x_text()

pg_sc <- ggscatter(result_Noiseq, x="Bud_mean", y="Cap_mean",color = "prob")+ 
    scale_x_log10()+scale_y_log10()

pg+pg_sc
```

<!-- Note that the results="hide" and fig.keep = "none" parameters were added to the code chunk to prevent printing of the R code results and the exploratory plots created in the code chunk. -->

### Available methods

Following implementations of popular methods are available here.
```{r warning=FALSE,results="hide", fig.keep = "none"}
# limma 
?use_limma_trend(se, colData_id, control, treatment, rank = FALSE, ...)
?use_limma_voom(se, colData_id, control, treatment, rank = FALSE, ...)

# edgeR 
?use_edgeR_exact(se, colData_id, control, treatment, rank = FALSE, ...)
?use_edgeR_GLM(se, colData_id, control, treatment, rank = FALSE, ...)

# deseq2
?use_deseq2(se, colData_id, control, treatment, rank = FALSE, ...)

# DELocal
?use_DELocal(se, colData_id, control, treatment, rank = FALSE, ...)

# noiseq
?use_NOIseq(se, colData_id, control, treatment, rank = FALSE, ...) 

# EBSeq 
?use_EBSeq(se, colData_id, control, treatment, rank = FALSE, ...)

# samseq
?use_SAMseq(se, colData_id, control, treatment, rank = FALSE, ...)
```

Advanced users can pass package specific arguments through '...' .

### Compare DE results

It is possible to execute all DE methods together and get aggregated results in a data.frame.
`broadSeq::use_multDE()` should be used for it.
<!-- # bayseq can not be installed. It is is deprecated. -->
```{r warning=FALSE,results="hide", fig.keep = "none"}
# First define a named list of functions
funs <- list(limma_trend = use_limma_trend, limma_voom = use_limma_voom,
             edgeR_exact = use_edgeR_exact, edgeR_glm = use_edgeR_GLM,
             deseq2 = use_deseq2, 
             DELocal = use_DELocal, noiseq = use_NOIseq, 
             EBSeq = use_EBSeq) 


multi_result <- broadSeq::use_multDE(
    se = se, 
    deFun_list = funs, return.df = TRUE,  
    colData_id = "stage", control = "Bud", treatment = "Cap", 
    rank = TRUE)
```

The column names of the resultant data.frame are prefixed with corresponding function names. 

```{r warning=FALSE}
head(multi_result)
# nrow(multi_result) == nrow(se)
colnames(multi_result)
```

#### Similarity of methods

DE methods are clustered based on their ranking of genes

<!-- # saveRDS(multi_result,"multi_result.rds") -->
```{r warning=FALSE,fig.width=6,fig.height=6,eval=TRUE}
# clusters <- multi_result %>% dplyr::select(ends_with("rank")) %>% arrange(DELocal_rank) %>% t() %>% dist() %>% hclust()
clusters <- multi_result %>% dplyr::select(ends_with("rank")) %>% t() %>% dist() %>% hclust()
plot(clusters,main =  "distance: Euclidean")

clusters <- multi_result %>% dplyr::select(ends_with("rank")) %>% arrange(DELocal_rank) %>% t() %>% dist(method = "maximum") %>% hclust()
plot(clusters,main =  "distance: maximum")

```

### Plots

#### Volcano 

Up regulated genes should be red or hot colors  

```{r warning=FALSE,fig.width=6,fig.height=6,eval=TRUE}
multi_result %>% broadSeq::volcanoPlot(
    pValName = "deseq2_padj",
    lFCName = "deseq2_log2FoldChange",
    labelName = "symbol",
    palette = c("purple","orange","grey"),
    selectedLabel =
        multi_result %>% dplyr::arrange(deseq2_padj) %>% pull(symbol) %>% head()
)

multi_result %>% broadSeq::volcanoPlot(
    pValName = "deseq2_padj",
    lFCName = "deseq2_log2FoldChange",
    labelName = "symbol",
    palette = "lancet" ,
    selectedLabel = list(criteria = "(`x` > 2 | `x` < -2) & (`y` > 10)")
) 
```

<!-- #### Boxplot -->
```{r warning=FALSE,fig.width=6,fig.height=6,eval=FALSE,echo=FALSE}
multi_result %>% dplyr::arrange(deseq2_padj) %>% pull(ensembl_gene_id) %>% head()
genes_plot(se, features = c("ENSMUSG00000035095","ENSMUSG00000037217"),
           x = "condition", y = "counts", add = "dotplot",palette = "npg")
genes_plot(se, features = c("ENSMUSG00000035095","ENSMUSG00000037217"),
           x = "condition", y = "counts", fill="condition", add = "jitter",palette = "npg")
```


```{r warning=FALSE,fig.width=6,fig.height=6,eval=FALSE,echo=FALSE}
# distinct_ranks <- which(cutree(clusters, k = 6) >1) %>% names()
distinct_ranks <- c("DELocal_rank", "limma_voom_rank","edgeR_glm_rank","EBSeq_rank") #,"bayseq_rank"
other_cols <- c("limma_voom_logFC","limma_voom_adj.P.Val","DELocal_adj.P.Val","limma_voom_AveExpr")

multi_result$ENSEMBL <- row.names(multi_result)


multi_result <- multi_result %>% mutate(DE_method = 
                                          ifelse((limma_voom_adj.P.Val < 0.05 & DELocal_adj.P.Val < 0.05), "Both",
                                                 ifelse(limma_voom_adj.P.Val < 0.05, "limma",
                                                        ifelse(DELocal_adj.P.Val < 0.05, "DELocal",
                                                               NA))) )
multi_result$DE_method <- factor(multi_result$DE_method,levels = c("DELocal","Both","limma",NA))
multi_result %>% dplyr::select(c(other_cols,distinct_ranks)) %>% 
  arrange(DELocal_rank) %>% head()

only_delocal_gene <- multi_result %>% dplyr::filter(DE_method=="DELocal") %>% 
    arrange(DELocal_rank) %>% pull(ensembl_gene_id)

# needs to implement using ggpubr
DELocal::plotNeighbourhood(pse = se,pDesign = ~ condition, 
                                 pGene_id = only_delocal_gene[1])$plot + 
  labs(title = names(only_delocal_gene[1]))

DELocal::plotNeighbourhood(pse = se,pDesign = ~ condition, 
                                 pGene_id = only_delocal_gene[2])$plot + 
  labs(title = names(only_delocal_gene[2]))

DELocal::plotNeighbourhood(pse = se,pDesign = ~ condition, 
                                 pGene_id = only_delocal_gene[3])$plot + 
  labs(title = names(only_delocal_gene[3]))


```


```{r echo=FALSE,warning=FALSE,message=FALSE,eval=FALSE}
multi_result[high_dev_gene,] %>% dplyr::select(c(other_cols,distinct_ranks)) %>% 
  arrange(DELocal_rank) %>%
  write.csv(file = "multi_result_selected_col.csv",row.names = T)

top_delocal_high_dev_gene <- multi_result[high_dev_gene,] %>% dplyr::select(ends_with("rank")) %>% 
  arrange(DELocal_rank) %>% rownames() 

for (gene in head(top_delocal_high_dev_gene)) {
 DELocal::plotNeighbourhood(pse = se,pDesign = ~ condition, pGene_id = gene)$plot %>% print()
}

pdf("high_devolopmental_genePlots.pdf")
for (gene in top_delocal_high_dev_gene) {
  # x <- multi_result[gene,] %>% dplyr::select(c(other_cols,distinct_ranks)) %>% round(digits = 4)
  # paste(names(x), x, sep = ":", collapse = "\n")
 DELocal::plotNeighbourhood(pse = se,pDesign = ~ condition, pGene_id = gene)$plot %>% print()
}
dev.off()
```

```{r echo=FALSE,warning=FALSE,eval=FALSE}
delocal_dev_choice <- rnaSeq_median_xprsn_devGenes %>% 
  dplyr::filter(ENSEMBL %in% rownames(DELocal_result %>% filter(adj.P.Val < 0.05)) & 
                  (rnaSeq_median_E14_Tooth > E14_Tooth_75_percentile | 
                     rnaSeq_median_E13_Tooth > E13_Tooth_75_percentile)) 

DELocal_result[rnaSeq_median_xprsn_devGenes$ENSEMBL,] %>% arrange(rank) %>% head()

rnaSeq_median_xprsn %>% filter(ENSEMBL=="ENSMUSG00000024598")
gPlot <- DELocal:::plotNeighbourhood(se,pDesign = ~ condition, pGene_id = "ENSMUSG00000024598")

rnaSeq_median_xprsn_devGenes %>% ggplot(aes(rnaSeq_median_E13_Tooth,rnaSeq_median_E14_Tooth,label=SYMBOL)) + 
  geom_point()+ 
  geom_hline(yintercept = E14_Tooth_75_percentile,)+ geom_vline(xintercept = E13_Tooth_75_percentile)+
  ggrepel::geom_text_repel(
    data = delocal_dev_choice,
    nudge_y = 2,
    segment.size  = 0.2,
    segment.color = "grey50",
    direction     = "x"
)
```


<details>
<summary>sessionInfo</summary>

```{r echo=FALSE,message=FALSE,warning=FALSE,error=FALSE}
sessionInfo()
```
</details>
