---
title: "Quality checking of experiment"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Quality checking of experiment}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, include = FALSE}
library(broadSeq)
library(ggplot2)
```

## Reading the data

Here we read gene expression data which is quantified by 'salmon' and
processed by `tximport` package. The 'salmon' generates count data, "abundance"
(TPM) and gene length values. Therefore we may rename the corresponding
assay slot 'abundance' to 'TPM'.

These data belongs to this studies reference!!

In this data, there are gene expression values from first molar tissue
of three different developing rodent species. The tissues are also from
four different developmental time points.

```{r message=FALSE}
se <- readRDS(system.file("extdata","rat_vole_mouseSE_salmon.rds", package = "broadSeq"))
SummarizedExperiment::assayNames(se)
```

### Sample metadata

```{r }
as.data.frame(colData(se)) %>% dplyr::count(stage,species) %>% tidyr::spread(stage,n)

se$stage <- factor(se$stage,levels = c("Bud","Cap","Late Cap","Bell"))
```

### Filtering out low expression genes

```{r lowExpr,warning=FALSE,fig.width= 6}
# assays(se)[["counts"]][,1] %>% gghistogram() +scale_x_log10()
assays(se)[["counts"]][,5] %>% ggpubr::ggdensity(y = "count")+
    ggplot2::geom_vline(xintercept = 10)+ggplot2::scale_x_log10()

keep <- (assays(se)[["counts"]] >= 3) %>% rowSums() >= 5 
# smallest Group Size is 5
table(keep)
```

## Normalization

edgeR provides two different methods ; CPM and TMM. But it does not use
normalized values for Differential Expression.

It also does not normalizes count values rather it normalizes the
library sizes using "TMM" method in 'normLibSizes'. Either use or not
use normLibSizes(). edgeR::cpm() will generate normalized expression
values.

### CPM

```{r warning=FALSE,message=FALSE}
se <- broadSeq::normalizeEdgerCPM(se ,method = "none",cpm.log = TRUE )
## The normalized values are added with the assay name "logCPM"
SummarizedExperiment::assayNames(se)
```

### TMM

<!-- 'There is no such thing as a "TMM normalized count" because TMM -->
<!-- normalizes the library sizes rather than the counts' -->
<!-- [see](https://www.biostars.org/p/9475236/) from Gordon Smyth (edgeR -->
<!-- developer). -->

```{r warning=FALSE,message=FALSE}
se <- broadSeq::normalizeEdgerCPM(se , method = "TMM", cpm.log = FALSE )
## The normalized values are added with the assay name "TMM"
SummarizedExperiment::assayNames(se)
```

#### access

```{r }
assays(se)[["counts"]][1:5,1:5]
assays(se)[["TMM"]][1:5,1:5]
assays(se)[["logCPM"]][1:5,1:5]
```

## DESeq2 Transformation

DESeq2 provides three different transformations

### VST

variance stabilizing transformation (VST)

```{r warning=FALSE}
se <- broadSeq::transformDESeq2(se,method = "vst"  )
```

### Normalized counts transformation

```{r }
se <- broadSeq::transformDESeq2(se, method = "normTransform"  )
```

### rlog

regularized log

```{r eval=FALSE}
se <- broadSeq::transformDESeq2(se, method = "rlog")
```

```{r }
SummarizedExperiment::assayNames(se)
```

### Comparision

Boxplot of different transforms for each sample.

```{r warning=FALSE,fig.height=6,fig.width=8}
p <- broadSeq::sampleAssay_plot(se[, se$species=="Mouse" ], 
                           assayName = "counts", fill = "stage", 
                           yscale = "log2")+ rremove("x.text")

p1 <- broadSeq::sampleAssay_plot(se[, se$species=="Mouse"], 
                           assayName = "vst", fill = "stage")+ rremove("x.text")

p2 <- broadSeq::sampleAssay_plot(se[, se$species=="Mouse"], 
                           assayName = "TMM", fill = "stage", 
                           yscale = "log10")+ rremove("x.text")

p3 <- broadSeq::sampleAssay_plot(se[, se$species=="Mouse"], 
                           assayName = "logCPM", fill = "stage")+ rremove("x.text")


ggarrange(p,p1,p2,p3, common.legend = TRUE, labels = c("A","B","C"))
```

Plot standard deviations versus means expression

```{r}
if (requireNamespace("vsn", quietly = TRUE)) {
    library("vsn")
    x <- meanSdPlot(
        log2(assays(se[, se$species == "Rat"])[["counts"]]+1),
        plot = FALSE)
    print(x$gg +ggtitle(label = "log2(n+1) "))
    
    x <- meanSdPlot(
        assays(se[, se$species == "Rat"])[["vst"]],
        plot = FALSE)
    
    print(x$gg +ggtitle(label = "Vst"))
    
    x <- meanSdPlot(
        assays(se[, se$species == "Rat"])[["logCPM"]],
        plot = FALSE)
    print(x$gg + ggtitle(label = "logCPM"))
}    

```

## Visualization of gene Expression

```{r message=FALSE,fig.width=8,fig.height=6,warning=FALSE}
## Multiple assay of a single gene
broadSeq::assay_plot(se, feature = c("Shh"), 
           assayNames = c("counts","logCPM","vst","TMM"),
           x = "stage", fill="species", add="dotplot", palette = "npg")

## Expression of multiple genes from a single assay
broadSeq::genes_plot(se, 
                     features = c("Shh","Edar"), 
                     facet.by = "symbol",
                     x = "stage", assayName = "vst", fill="species", palette = "jco")
```

#### Pre-defined or custom color palette based on journals

Scientific journal palettes from `ggsci` R package, e.g.: "npg", "aaas",
"lancet", "jco", "ucscgb", "uchicago", "simpsons" and "rickandmorty" can
be passed for coloring or filling by groups from metadata.

```{r warning=FALSE,message=FALSE,fig.height=6,fig.width=8}
jco <- broadSeq::genes_plot(se[,se$species == "Mouse"], 
                     features = c("Shh"), facet.by = "symbol", assayName =  "logCPM",
                     x = "stage",  fill="stage", add="dotplot", xlab = "", 
                     title = "Journal of Clinical Oncology", palette = "jco") 

npg <- broadSeq::genes_plot(se[,se$species == "Mouse"], 
                     features = c("Shh"), facet.by = "symbol",assayName =  "logCPM",
                     x = "stage", fill="stage", add="dotplot", xlab = "",
                     title = "Nature Publishing Group", palette = "npg") 

aaas <- broadSeq::genes_plot(se[,se$species == "Mouse"], 
                     features = c("Shh"), facet.by = "symbol", assayName = "logCPM",
                     x = "stage", fill="stage", add="dotplot", xlab = "",
                     title = "Science", palette = "aaas")

nejm <- broadSeq::genes_plot(se[,se$species == "Mouse"], 
                     features = c("Shh"), facet.by = "symbol", assayName = "logCPM",
                     x = "stage", fill="stage", add="dotplot",  xlab = "",
                     title = "New England Journal of Medicine",palette = "nejm")

# ggarrange(jco,npg,aaas,nejm, 
#           common.legend = TRUE,legend = "none",
#           labels = c("A","B","C","D"))

ggarrange(jco+ggpubr::rotate_x_text(), npg+ggpubr::rotate_x_text(),
          aaas+ggpubr::rotate_x_text(),nejm+ggpubr::rotate_x_text(),
          nrow = 1, common.legend = TRUE,legend = "none",
          labels = c("A","B","C","D")) %>% 
    annotate_figure( top = text_grob("Color palette")) 
```

## QC with Clustering

```{r}
## Number of top variable genes
ntop=500
## How it matters?
```

### MDS plot

Classical multidimensional scaling is based on measuring the distance
between the samples.

Popular function plotMDS from `limma` does not work with
`SummarizedExperiment` object. Here broadSeq provides this function
through package ´cmdscale {stats}´.

[Classical multidimensional scaling (MDS) of a data matrix. Also known
as principal coordinates analysis (Gower, 1966). ]

```{r fig.height=6,fig.width=8}
broadSeq::plot_MDS(se, scaledAssay = "vst", ntop=500, 
                   color = "species", shape = "stage", 
                   ellipse=TRUE, legend = "bottom")
```

MDS is cool but it is not possible to know/visualize top variable genes
with their meta data which is stored in `se` object

```{r }
head(rowData(se))
```

Other methods can help to visualize gene information along with
clustering information.

### Hierarchical clustering and Heatmap

```{r fig.height=6,fig.width=8}
p_vst <- broadSeq::plotHeatmapCluster(
    se,
    scaledAssay = "vst",
    annotation_col = c("species", "stage"),
    annotation_row = c("Class","gene_biotype"),
    ntop = 30, show_geneAs = "symbol",
    cluster_cols = TRUE, cluster_rows = FALSE,
    show_rownames = TRUE, show_colnames = FALSE,
    main = "Top 30 variable gene vst"
)

```

### PCA plot

#### prcompTidy

Perform Principal Components Analysis with function
`broadSeq::prcompTidy()` which returns a list of four `data.frame`
objects:

-   pc_scores,

-   eigen_values,

-   loadings (eigen vectors) and

-   the original data.

    Compute PCA using any assay

```{r warning=FALSE}
computedPCA_logCPM <- broadSeq::prcompTidy(se, scaledAssay = "logCPM", ntop = 500)
## PCA based on vst values
computedPCA_vst <- broadSeq::prcompTidy(se, scaledAssay = "vst", ntop = 500)
```

#### Plot

##### logCPM

```{r fig.width=8,fig.height=6}
plotAnyPC(computedPCA = computedPCA_logCPM,
          x = 1, y = 2, color = "species", shape = "stage",
          legend = "bottom")
```

##### VST

```{r fig.width=8,fig.height=6}
pca_vst <- plotAnyPC(computedPCA = computedPCA_vst,
            x = 1, y = 2,  color = "species", shape = "stage", 
            legend = "bottom") 
pca_vst
```

##### Other PCs

```{r fig.width=8,fig.height=6}
computedPCA_vst$eigen_values %>%
        dplyr::filter(var_exp >= 2) %>%
    ggbarplot(x="PC",y="var_exp", label = TRUE, label.pos = "out")
```

It can be checked if there are other PCs to explain considerable
variance. PC3 can be useful to see variance due to different
developmental time points.

```{r fig.width=8,fig.height=6}
pca_vst_2_3 <-plotAnyPC(computedPCA = computedPCA_vst,
                x = 2, y = 3,  
                color = "species", shape = "stage", legend = "bottom")
# pca_vst_2_3
```

PC3 captures beautifully the variance in gene expression due to
developmental stages.
<!-- The PCAs are different since the top 500 genes are different for rlog and vst.   -->

#### Gene loading

```{r fig.width=7,fig.height=6}
computedPCA_vst %>% broadSeq::getFeatureLoadRanking(keep = c("symbol","Class")) %>% head()

#  Top 5 genes of PC2

computedPCA_vst$loadings %>% top_n(5,abs(PC2)  ) %>% dplyr::select(gene,PC2)

pca_vst_loading <- computedPCA_vst %>% 
    broadSeq::getFeatureLoadRanking(keep = c("symbol","Class"), topN = 50, pcs=1:10) %>% 
    dplyr::count(Class, PC) %>%
    ggbarplot(
        x = "PC", y = "n", fill = "Class",
        legend = "bottom", palette = c("red","blue","orange","purple","white","grey")
    ) 
# pca_vst_loading
```

#### Biplot

```{r fig.width=8,fig.height=6}
# By default it plots top 2 genes from each PC axis
pca_vst_bi <- broadSeq::biplotAnyPC(computedPCA = computedPCA_vst, 
            x = 1, y = 2, genesLabel = "symbol", 
            color = "species", shape = "stage", 
            legend = "bottom")
# pca_vst_bi
```


```{r fig.width=8,fig.height=8}
ggarrange(
    ggarrange(pca_vst_bi+ggtitle(label =  ""),
          pca_vst_2_3+ggtitle(label =  ""), common.legend = TRUE),
    pca_vst_loading, nrow = 2)

```

##### To plot user defined genes

Now plotting top 5 genes from PC3

```{r fig.width=8,fig.height=6}
# Top 5 genes of PC3
biplotAnyPC(computedPCA = computedPCA_vst,x = 2, y = 3, 
            color = "species", shape = "stage",
            genes= computedPCA_vst$loadings %>% 
                top_n(5,abs(PC3)) %>% pull(gene),
            genesLabel = "symbol")

## Plot progression gene "Shh" 
biplotAnyPC(computedPCA = computedPCA_vst,x = 2, y = 3, 
            color = "species", shape = "stage",
            genes=c("Shh"),
            genesLabel = "symbol")
```

<details>

<summary>sessionInfo</summary>

```{r echo=FALSE,message=FALSE,warning=FALSE,error=FALSE}
sessionInfo()
```

</details>
